beacon-chain/state/stategen/history.go

```diff
@@ -0,0 +1,205 @@
+package stategen
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	types "github.com/prysmaticlabs/eth2-types"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	"github.com/prysmaticlabs/prysm/beacon-chain/db"
+	"github.com/prysmaticlabs/prysm/beacon-chain/state"
+	"github.com/prysmaticlabs/prysm/encoding/bytesutil"
+	"github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1/block"
+	"go.opencensus.io/trace"
+)
+
+func WithCache(c CachedGetter) CanonicalHistoryOption {
+	return func(h *CanonicalHistory) {
+		h.cache = c
+	}
+}
+
+type CanonicalHistoryOption func(*CanonicalHistory)
+
+func NewCanonicalHistory(h HistoryAccessor, cc CanonicalChecker, cs CurrentSlotter, opts ...CanonicalHistoryOption) *CanonicalHistory {
+	ch := &CanonicalHistory{
+		h:  h,
+		cc: cc,
+		cs: cs,
+	}
+	for _, o := range opts {
+		o(ch)
+	}
+	return ch
+}
+
+type CanonicalHistory struct {
+	h     HistoryAccessor
+	cc    CanonicalChecker
+	cs    CurrentSlotter
+	cache CachedGetter
+}
+
+func (ch *CanonicalHistory) ReplayerForSlot(target types.Slot) Replayer {
+	return &stateReplayer{chainer: ch, method: forSlot, target: target}
+}
+
+func (c *CanonicalHistory) BlockForSlot(ctx context.Context, target types.Slot) ([32]byte, block.SignedBeaconBlock, error) {
+	currentSlot := c.cs.CurrentSlot()
+	if target > currentSlot {
+		return [32]byte{}, nil, errors.Wrap(ErrFutureSlotRequested, fmt.Sprintf("requested=%d, current=%d", target, currentSlot))
+	}
+	for target > 0 {
+		if ctx.Err() != nil {
+			return [32]byte{}, nil, errors.Wrap(ctx.Err(), "context canceled during canonicalBlockForSlot")
+		}
+		hbs, err := c.h.HighestSlotBlocksBelow(ctx, target+1)
+		if err != nil {
+			return [32]byte{}, nil, errors.Wrap(err, fmt.Sprintf("error finding highest block w/ slot <= %d", target))
+		}
+		if len(hbs) == 0 {
+			return [32]byte{}, nil, errors.Wrap(ErrNoBlocksBelowSlot, fmt.Sprintf("slot=%d", target))
+		}
+		r, b, err := c.bestForSlot(ctx, hbs)
+		if err == nil {
+			// we found a valid, canonical block!
+			return r, b, nil
+		}
+
+		// we found a block, but it wasn't considered canonical - keep looking
+		if errors.Is(err, ErrNoCanonicalBlockForSlot) {
+			// break once we've seen slot 0 (and prevent underflow)
+			if hbs[0].Block().Slot() == 0 {
+				break
+			}
+			target = hbs[0].Block().Slot() - 1
+			continue
```

Comment:
I don't think we need this. The way I understand `HighestSlotBlocksBelow` is it would just return you the highest slot block no matter what

Reply:
The loop is handling the case where `HighestSlotBlocksBelow` does not give us a canonical block. In that case we decrement the slot and keep searching backwards.

Reply:
This is using `IsCanonical` from fork choice store.

Reply:
I see. Great to see we are covering all the edge cases!